x/observer/keeper/msg_server_add_block_header.go

```diff
@@ -35,6 +34,38 @@ func (k msgServer) AddBlockHeader(goCtx context.Context, msg *types.MsgAddBlockH
 		return nil, cosmoserrors.Wrapf(types.ErrBlockHeaderVerficationDisabled, "proof verification not enabled for evm ,chain id: %d", msg.ChainId)
 	}
 
+	_, found = k.GetBlockHeader(ctx, msg.BlockHash)
+	if found {
+		return nil, cosmoserrors.Wrap(types.ErrBlockAlreadyExist, "cannot add block header")
+	}
+
+	bhs, found := k.Keeper.GetBlockHeaderState(ctx, msg.ChainId)
+	if found && bhs.EarliestHeight > 0 && bhs.EarliestHeight < msg.Height {
```

Comment:
We should add comments to explain this condition

Reply:
done [d8eacc3](https://github.com/zeta-chain/node/pull/1374/commits/d8eacc36e3f6c2e6ec2242a8cf4a486f9f38c489)